### ISSUE TITLE
Correctly handle `F.interpolate` upsample with amp

### ISSUE THREAD

**Issue by bhack** [pytorch/pytorch#121072]
### 🐛 Describe the bug

https://github.com/pytorch/pytorch/issues/104157

As this bug seems to be still a quite tricky  for many user can we let correctly handle this case when using amp?

I see developer are using manual tricks to workaround this bug:
https://github.com/huggingface/transformers/pull/28504/files#r1455033425

But as often amp is automated on the whole model it could be insidious for many user.

Can amp automatically handle this case internally? 

### Versions

main/nightly

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @mcarilli @ptrblck @leslie-fang-intel @jgong5

**Comment by bhack** [comment]
Also is that bug still valid with `bfloat16`?

**Comment by ezyang** [comment]
It seems like it should be relatively simple to impede ALP on these ill behaved ops. Do you think you could try patching it?

**Comment by bhack** [comment]
Is bfloat16 involved?

**Comment by ezyang** [comment]
I'm not sure, you'll have to check

**Comment by bhack** [comment]
Expanding the original ticket code
```python
import torch
from torch.nn.functional import interpolate

def get_grads(dtype, size):
    input = torch.randn((2, 2, 1, 1), dtype=dtype, requires_grad=True).cuda()

    output = interpolate(input, size=size, mode='bicubic', align_corners=True)

    grad_outputs = [torch.ones_like(output)]
    grads = torch.autograd.grad([output], [input], grad_outputs)
    return grads[0]

print(get_grads(torch.float32, (128, 128)).flatten())
print(get_grads(torch.float16, (128, 128)).flatten())
print(get_grads(torch.bfloat16, (128, 128)).flatten())

print(get_grads(torch.float32, (64, 64)).flatten())
print(get_grads(torch.float16, (64, 64)).flatten())
print(get_grads(torch.bfloat16, (64, 64)).flatten())

print(get_grads(torch.float32, (32, 32)).flatten())
print(get_grads(torch.float16, (32, 32)).flatten())
print(get_grads(torch.bfloat16, (32, 32)).flatten())
````

```python
tensor([16384., 16384., 16384., 16384.], device='cuda:0')
tensor([2048., 2048., 2048., 2048.], device='cuda:0', dtype=torch.float16)
tensor([256., 256., 256., 256.], device='cuda:0', dtype=torch.bfloat16)
tensor([4096., 4096., 4096., 4096.], device='cuda:0')
tensor([2048., 2048., 2048., 2048.], device='cuda:0', dtype=torch.float16)
tensor([256., 256., 256., 256.], device='cuda:0', dtype=torch.bfloat16)
tensor([1024., 1024., 1024., 1024.], device='cuda:0')
tensor([1024., 1024., 1024., 1024.], device='cuda:0', dtype=torch.float16)
tensor([256., 256., 256., 256.], device='cuda:0', dtype=torch.bfloat16)
```

**Comment by bhack** [comment]
What do you think it could be the best action?

**Comment by ezyang** [comment]
So in aten/src/ATen/autocast_mode.cpp there's a list of special rules, and the idea would be to add upsample to the list with the specific autocast behavior it needs.

**Comment by bhack** [comment]
I don't see `upsample` in https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/autocast_mode.cpp. So when it isn't there how it is handled?

**Comment by albanD** [comment]
You can see the registration here https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/autocast_mode.cpp#L244-L246 which is: does nothing (whatever the input dtype is is used).

if you want to force it to be fp32, you want to add an entry here: https://github.com/pytorch/pytorch/blob/7feabe9b73e6ba7724b62ea91df27049defdf378/aten/src/ATen/autocast_mode.cpp#L286

**Comment by bhack** [comment]
@albanD Something like https://github.com/pytorch/pytorch/pull/121324 ?

**Comment by albanD** [comment]
If the conclusion of the discussion here that we should force fp32, then yes that is the way to go

**Comment by vfdev-5** [comment]
I would say there can be potentially an issue with gradients if interpolate is used in a model with amp. Here is a repro code and the output + explicit printing of input/output and grad_input/grad_output dtypes from aten impl:
```python

import torch
import torch.nn as nn


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.m = nn.Conv2d(3, 3, 3, padding=1, bias=False)

    def forward(self, x):
        print("1 x dtype", x.dtype)
        x = self.m(x)
        print("2 x dtype", x.dtype)
        x = nn.functional.interpolate(x, scale_factor=(2.5, 2.5), mode="bicubic", align_corners=False)
        print("3 x dtype", x.dtype)
        return x

def func(enabled, dtype=torch.float16):
    torch.manual_seed(12)
    model = Model().cuda()
    x = torch.rand(2, 3, 32, 32, device="cuda")

    with torch.cuda.amp.autocast(enabled=enabled, dtype=dtype):
        out = model(x)
        print(out.dtype)
        loss = out.mean()

    loss.backward()
    grad = model.m.weight.grad
    print(grad.dtype, grad.max(), grad.min(), grad.mean())


print("\n--- No AMP")
func(False)
print("\n--- AMP using f16")
func(True, dtype=torch.float16)
print("\n--- AMP using bf16")
func(True, dtype=torch.bfloat16)
```
Output:
```

--- No AMP
1 x dtype torch.float32
2 x dtype torch.float32
FWD input dtype: float
FWD output dtype: float
3 x dtype torch.float32
torch.float32
BWD grad_output dtype: float
BWD grad_input dtype: float
torch.float32 tensor(0.1684, device='cuda:0') tensor(0.1543, device='cuda:0') tensor(0.1597, device='cuda:0')

--- AMP using f16
1 x dtype torch.float32
2 x dtype torch.float16
FWD input dtype: c10::Half
FWD output dtype: c10::Half
3 x dtype torch.float16
torch.float16
BWD grad_output dtype: c10::Half
BWD grad_input dtype: c10::Half
torch.float32 tensor(0.1688, device='cuda:0') tensor(0.1547, device='cuda:0') tensor(0.1600, device='cuda:0')

--- AMP using bf16
1 x dtype torch.float32
2 x dtype torch.bfloat16
FWD input dtype: c10::BFloat16
FWD output dtype: c10::BFloat16
3 x dtype torch.bfloat16
torch.bfloat16
BWD grad_output dtype: c10::BFloat16
BWD grad_input dtype: c10::BFloat16
torch.float32 tensor(0.1670, device='cuda:0') tensor(0.1533, device='cuda:0') tensor(0.1582, device='cuda:0')
```

If put `KERNEL_CUDA(upsample_bicubic2d, fp32)` into `TORCH_LIBRARY_IMPL(aten, Autocast, m) {`, pytorch/aten/src/ATen/autocast_mode.cpp then AMP uses float32 in forward and backward passes.
The same code gives the following output:
```
--- No AMP
1 x dtype torch.float32
2 x dtype torch.float32
FWD input dtype: float
FWD output dtype: float
3 x dtype torch.float32
torch.float32
BWD grad_output dtype: float
BWD grad_input dtype: float
torch.float32 tensor(0.1684, device='cuda:0') tensor(0.1543, device='cuda:0') tensor(0.1597, device='cuda:0')

--- AMP using f16
1 x dtype torch.float32
2 x dtype torch.float16
FWD input dtype: float
FWD output dtype: float
3 x dtype torch.float32
torch.float32
BWD grad_output dtype: float
BWD grad_input dtype: float
torch.float32 tensor(0.1685, device='cuda:0') tensor(0.1543, device='cuda:0') tensor(0.1597, device='cuda:0')

--- AMP using bf16
1 x dtype torch.float32
2 x dtype torch.bfloat16
FWD input dtype: float
FWD output dtype: float
3 x dtype torch.float32
torch.float32
BWD grad_output dtype: float
BWD grad_input dtype: float
torch.float32 tensor(0.1689, device='cuda:0') tensor(0.1543, device='cuda:0') tensor(0.1599, device='cuda:0')
```

However, my question here would be how often we have large gradients (as in the example: https://github.com/pytorch/pytorch/issues/121072#issuecomment-1979083131) such that they can't be represented by float16 dtype. Should we also use gradscaler in such cases ?

**Comment by bhack** [comment]
> However, my question here would be how often we have large gradients (as in the example:

Do we have a pool of real models using upsample (e.g. segmentation netwrok etc..) e.g. from Huggingface that we can test with the CI or with an internal CI job?

**Comment by bhack** [comment]
The focus could be more about a pyramid decoder (e.g. segmentation) where generally we could have a chain of upsampling layers.
